sdk/python/kfp/cli/compile_.py

```diff
@@ -22,36 +22,67 @@
 
 import click
 from kfp import compiler
-from kfp.components import pipeline_context
-from kfp.components import python_component
+from kfp.components import base_component
+from kfp.components import graph_component
 
 
-def collect_pipeline_from_module(
+def is_pipeline_func(func: Callable) -> bool:
+    """Checks if a function is a pipeline function.
+
+    Args:
+        func (Callable): The function to check.
+
+    Returns:
+        bool: True if the function is a pipeline function.
+    """
+    return isinstance(func, graph_component.GraphComponent)
+
+
+def is_component_func(func: Callable) -> bool:
+    """Checks if a function is a component function.
+
+    Args:
+        func (Callable): The function to check.
+
+    Returns:
+        bool: True if the function is a component function.
+    """
+    return not is_pipeline_func(func) and isinstance(
+        func, base_component.BaseComponent)
+
+
+def collect_pipeline_or_component_from_module(
     target_module: types.ModuleType
-) -> Union[Callable[..., Any], python_component.PythonComponent]:
+) -> Union[Callable[..., Any], base_component.BaseComponent]:
     pipelines = []
+    components = []
     module_attrs = dir(target_module)
     for attr in module_attrs:
         obj = getattr(target_module, attr)
-        if pipeline_context.Pipeline.is_pipeline_func(obj):
+        if is_pipeline_func(obj):
             pipelines.append(obj)
+        elif is_component_func(obj):
+            components.append(obj)
+
     if len(pipelines) == 1:
         return pipelines[0]
+    elif len(pipelines) == 0 and len(components) == 1:
+        return components[0]
     else:
         raise ValueError(
-            f'Expect one pipeline function in module {target_module}, got {len(pipelines)}: {pipelines}. Please specify the pipeline function name with --function.'
+            f'Expect one pipeline or component function in module {target_module}, got {len(pipelines)}: {pipelines}. Please specify the pipeline or component function name with --function.'
```

Comment:
Not sure what would be a better message here. If users compile a file containing multiple component functions and one pipeline function, they won't be asked to specify `--function`. This change reserves the behavior so it's not a breaking change. If it were for the breaking change concern, ideally we should ask users to specify `--function` whenever there're multiple pipeline/component functions.

Reply:
A few ideas:
- Unpacking the `one pipeline` and `one component` into separate terms in the message might help clarify the underlying logic a bit
- the output right now isn't particularly readable, since it prints the `__str__` of the object; perhaps we can push the reader in the right direction via `[p.name for p in pipelines]` and same for components
- nit: can replace `pipeline(s)` with `pipelines`, since pipelines count is necessarily >1 (this was probably my code :) )

All together, maybe something like:
`f'Expected one pipeline or one component in module {target_module}. Got {len(pipelines} {[p.name for p in pipelines]} and {len(components)} {[c.name for c in components]}. Please specify which pipeline or component to compile using --function.`

Reply:
> nit: can replace pipeline(s) with pipelines, since pipelines count is necessarily >1 (this was probably my code :) )

With the additional support for components, it could be 0 pipeline and >1 components.
Overall, I like your suggestions. Updated.